File: Data Structures & Algorithms/course-schedule/submission-2.py
from collections import defaultdict

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graphs = defaultdict(list)

        for u, v in prerequisites:
            graphs[u].append(v)

        UNVISITED, VISITING, VISITED = 0, 1, 2
        track = [0] * numCourses

        def dfs(node):
            if track[node] == VISITING:
                return False
            
            elif track[node] == VISITED:
                return True

            else:
                track[node] = VISITING

                for v in graphs[node]:
                    if not dfs(v):
                        return False

            track[node] = VISITED
            return True

        for i in range(numCourses):
            if not dfs(i):
                return False

        return True

File: Data Structures & Algorithms/course-schedule-ii/submission-2.go
func findOrder(numCourses int, prerequisites [][]int) []int {
	const (
        UNVISITED = 0
        VISITING  = 1
        VISITED   = 2
    )

    graph := make([][]int, numCourses)
    for _, p := range prerequisites {
        u, v := p[0], p[1]
        graph[u] = append(graph[u], v)
    }

    track := make([]int, numCourses)
	res := []int {}

    var dfs func(node int) bool
    dfs = func(node int) bool {
        if track[node] == VISITING {
            return false
        }
        if track[node] == VISITED {
            return true
        }

        track[node] = VISITING
        for _, pre := range graph[node] {
            if !dfs(pre) {
                return false
            }
        }
        track[node] = VISITED
		res = append(res, node)
        return true
    }

    for i := 0; i < numCourses; i++ {
        if !dfs(i) {
            return []int {}
        }
    }

    return res
}
